src/compiler/builder.ts

```diff
@@ -433,7 +433,9 @@ namespace ts {
             return;
         }
 
-        forEachReferencingModulesOfExportOfAffectedFile(state, affectedFile, (state, path) => handleDtsMayChangeOf(state, path, cancellationToken, computeHash));
+        if (!state.compilerOptions.assumeChangesOnlyAffectDirectDependencies) {

```

Comment:
To someone unfamiliar with the builder, it's not clear how adding this simple check does what it says in the description.  Where are the direct dependencies handled if not in the call below?  At what point is the list expanded to include indirect dependencies?

Reply:
getAffectedFiles handle the direct dependency and the below function handles indirect ones.